worker/raft/raftclusterer/worker.go

```diff
@@ -0,0 +1,263 @@
+	prevIndex uint64,
+	details apiserver.Details,
+) (uint64, error) {
+	newServers := make(map[raft.ServerID]raft.ServerAddress)
+	for _, server := range details.Servers {
+		if server.InternalAddress == "" {
+			continue
+		}
+		serverID := raft.ServerID(names.NewMachineTag(server.ID).String())
+		serverAddress := raft.ServerAddress(server.InternalAddress)
+		newServers[serverID] = serverAddress
+	}
+	if len(newServers) < 3 {
+		// NOTE(axw) when we bootstrap the cluster, we set
+		// the first node's address to be localhost. Until
+		// the cluster is grown beyond a single node, we
+		// can never update the address.
+		//
+		// Changing a node's address can only be done by
+		// removing it from the configuration and adding it
+		// back in with its new address; this cannot be
+		// done if there is only one node, as there must
+		// be one voting node in the configuration at all
+		// times.
+		return prevIndex, nil
+	}
+	ops := w.configOps(configuredServers, newServers)
+	for _, op := range ops {
+		future := op(prevIndex)
+		if err := future.Error(); err != nil {
+			return 0, errors.Trace(err)
+		}
+		prevIndex = future.Index()
+	}
+	return prevIndex, nil
+}
+
+func (w *Worker) configOps(configuredServers, newServers map[raft.ServerID]raft.ServerAddress) []configOp {
+	// This worker can only run on the raft leader, ergo Leader
+	// returns the local address.
+	localAddr := w.config.Raft.Leader()
+
+	// Add new servers first, then remove old servers,
+	// and finish by changing addresses. If the local
+	// server is to be removed, we do that last, to
+	// ensure that the other servers can communicate
+	// amongst themselves.
+	var ops []configOp
+	for id, newAddr := range newServers {
+		if _, ok := configuredServers[id]; ok {
+			continue
+		}
+		logger.Infof("server %q added with address %q", id, newAddr)
+		ops = append(ops, w.addVoterOp(id, newAddr))
+		configuredServers[id] = newAddr
+	}
+	var removeLocal raft.ServerID
+	for id, oldAddr := range configuredServers {
+		if _, ok := newServers[id]; ok {
+			continue
+		}
+		logger.Infof("server %q removed", id)
+		delete(configuredServers, id)
+		if oldAddr == localAddr {
+			removeLocal = id
+			continue
+		}
+		ops = append(ops, w.removeServerOp(id))
+	}
+	for id, newAddr := range newServers {
+		oldAddr := configuredServers[id]
+		if oldAddr == newAddr {
+			continue
+		}
+		logger.Infof("server %q address changed from %q to %q", id, oldAddr, newAddr)
+		configuredServers[id] = newAddr
+		if oldAddr == localAddr {
+			removeLocal = id
+			continue
+		}
+		ops = append(ops, w.removeServerOp(id))
+		ops = append(ops, w.addVoterOp(id, newAddr))
+	}
+	if removeLocal != "" {
+		// Whether the local (leader) server was removed or
+		// its address has changed, here we simply remove it
+		// from the raft configuration. This will prompt
+		// another server to become leader; if the address
+		// changed, then the new leader will add the server
+		// back in with its new address.
```

Comment:
> AddVoter only promotes entries to be Staging. Should we be doing something about waiting for them to become fully voting before we remove the current leader? I suppose this would otherwise just generate an error of some sort (unable to remove leader?)

RemoveServer will fail if there are no other voting servers. We should probably wait for there to be a majority, though.

> If it is simply a changed address, doesn't that cause us to go through a bad period of not having an odd number of voters?

Indeed, that is a problem with this approach, none of which I particularly like. Perhaps have ID=address=machine-tag, but just map to a single internal/HA address as we're now doing. That ensures we communicate in the HA space, and eliminates the possibility of split-brain.

> Anyway, we can probably live with this, as address changes should be very rare.

This will *always* happen when you run enable-ha for the first time, so there is a chance that a split-brain could occur at that time.

Reply:
For posterity: AddVoter apparently can update the address of an existing server, you don't have to remove and add.